src/data/nat/digits.lean

```diff
@@ -0,0 +1,346 @@
+end
+
+lemma of_digits_modeq (b k : ℕ) (L : list ℕ) : (of_digits b L) ≡ (of_digits (b % k) L) [MOD k] :=
+of_digits_modeq' b (b % k) k (nat.modeq.symm (nat.modeq.mod_modeq b k)) L
+
+lemma of_digits_mod (b k : ℕ) (L : list ℕ) : (of_digits b L) % k = (of_digits (b % k) L) % k :=
+of_digits_modeq b k L
+
+lemma of_digits_zmodeq' (b b' : ℤ) (k : ℕ) (h : b ≡ b' [ZMOD k]) (L : list ℕ) : (of_digits b L) ≡ (of_digits b' L) [ZMOD k] :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [int.modeq] at *,
+    conv_lhs { rw [int.add_mod, int.mul_mod, h, L_ih], },
+    conv_rhs { rw [int.add_mod, int.mul_mod], }, }
+end
+
+lemma of_digits_zmodeq (b : ℤ) (k : ℕ) (L : list ℕ) : (of_digits b L) ≡ (of_digits (b % k) L) [ZMOD k] :=
+of_digits_zmodeq' b (b % k) k (int.modeq.symm (int.modeq.mod_modeq b ↑k)) L
+
+lemma of_digits_zmod (b : ℤ) (k : ℕ) (L : list ℕ) : (of_digits b L) % k = (of_digits (b % k) L) % k :=
+of_digits_zmodeq b k L
+
+lemma modeq_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+   n ≡ (digits b' n).sum [MOD b] :=
+begin
+  rw ←of_digits_one,
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  convert of_digits_modeq _ _ _,
+  exact h.symm,
+end
+
+lemma modeq_three_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 3] :=
+modeq_digits_sum 3 10 (by norm_num) n
+
+lemma modeq_nine_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 9] :=
+modeq_digits_sum 9 10 (by norm_num) n
+
+lemma dvd_iff_dvd_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+  b ∣ n ↔ b ∣ (digits b' n).sum :=
+begin
+  rw ←of_digits_one,
+  conv_lhs { rw ←(of_digits_digits b' n) },
+  rw [nat.dvd_iff_mod_eq_zero, nat.dvd_iff_mod_eq_zero, of_digits_mod, h],
+end
+
+lemma three_dvd_iff (n : ℕ) : 3 ∣ n ↔ 3 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 3 10 (by norm_num) n
+
+lemma nine_dvd_iff (n : ℕ) : 9 ∣ n ↔ 9 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 9 10 (by norm_num) n
+
+lemma zmodeq_of_digits_digits (b b' : ℕ) (c : ℤ) (h : b' ≡ c [ZMOD b]) (n : ℕ) :
+  n ≡ (of_digits c (digits b' n)) [ZMOD b] :=
+begin
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  rw coe_int_of_digits,
+  apply of_digits_zmodeq' _ _ _ h,
+end
+
+lemma modeq_eleven_digits_sum (n : ℕ) : n ≡ of_digits (-1 : ℤ) (digits 10 n) [ZMOD 11] :=
+zmodeq_of_digits_digits 11 10 (-1 : ℤ) dec_trivial n
+
+-- TODO move?
+lemma dvd_iff_dvd_of_dvd_sub {a b c : ℤ} (h : a ∣ (b - c)) : (a ∣ b ↔ a ∣ c) :=
+begin
+  split,
+  intro h',
+  convert dvd_sub h' h,
+  exact eq.symm (sub_sub_self b c),
+  intro h',
+  convert dvd_add h h',
+  exact eq_add_of_sub_eq rfl,
+end
+
+lemma dvd_iff_dvd_of_digits (b b' : ℕ) (c : ℤ) (h : (b : ℤ) ∣ (b' : ℤ) - c) (n : ℕ) :
+  b ∣ n ↔ (b : ℤ) ∣ (of_digits c (digits b' n)) :=
+begin
+  rw ←int.coe_nat_dvd,
+  exact dvd_iff_dvd_of_dvd_sub
+    (int.modeq.modeq_iff_dvd.1 (zmodeq_of_digits_digits b b' c (int.modeq.modeq_iff_dvd.2 h).symm _).symm),
+end
+
+lemma eleven_dvd_iff (n : ℕ) : 11 ∣ n ↔ 11 ∣ (of_digits (-1 : ℤ) (digits 10 n)) :=
```

Comment:
That's a goofy way of writing "alternating sum" :rofl:

Reply:
What do you think is the nicest way to express this? I could use `foldr`, we could define `list.alternating_sum`, I could write a summation, ...?

Reply:
I think `list.alternating_sum` makes sense. But the good old
```
\sum i in (range n), (-1 : int) ^ i * foobar
```
also looks recognisable.

Reply:
I made a definition of `alternating_sum` (but just left it locally in this file, as I don't prove anything about it).

I did try to prove 
```
lemma alternating_sum_eq_finset_sum {R : Type*} [ring R] (L : list R) :
  alternating_sum L =
    ∑ i : fin L.length, (-1 : ℤ) ^ (i : ℕ) * L.nth_le i i.2 :=
sorry
```
but my big-operator-fu is not strong tonight.

Reply:
Thanks, @jcommelin, for providing the proof!